crates/shrs_core/src/state.rs

```diff
@@ -78,31 +80,91 @@ impl<T: 'static> DerefMut for StateMut<'_, T> {
         self.value.downcast_mut().unwrap()
     }
 }
+
+// Potential errors that can occur when interacting with state store
+#[derive(Error, Debug)]
+pub enum StateError {
+    // TODO include the type in the error message
+    #[error("Value is missing")]
+    Missing,
+    #[error("Failed to borrow")]
+    Borrow,
+    #[error("Failed to borrow mut")]
+    BorrowMut,
+    #[error("Failed to downcast")]
+    Downcast,
+}
+
+// Global state store
 #[derive(Default)]
 pub struct States {
     states: HashMap<TypeId, RefCell<Box<dyn Any>>>,
 }
+
 impl States {
+
+    // Insert a new piece of state of given type into global state store, overriding previously
+    // existing values
+    // TODO should we allow overriding contents of state?
```

Comment:
@nithinmuthukumar thoughts?

Reply:
Currently inserting a state of the same type overrides the current state.